docs/src/api/tipg/settings.md
::: tipg.settings


docs/src/advanced/configuration.md

`tipg` uses [starlette](https://www.starlette.io/config/)'s configuration pattern which make use of environment variable and/or `.env` file.

Example of `.env` file can be found in [.env.example](https://github.com/developmentseed/tipg/blob/main/.env.example)

## DB connection settings

class: `tipg.settings.PostgresSettings`

#### Database address

You can pass either the Postgres DSN using `DATABASE_URL` or with each specific values for usernane, password, host, port and database name.

- **POSTGRES_USER** (str): Username
- **POSTGRES_PASS** (str): Password
- **POSTGRES_HOST** (str): Hostname (e.g `127.0.0.1`)
- **POSTGRES_PORT** (str): Port number (e.g `5432`)
- **POSTGRES_DBNAME** (str): Database name (e.g `postgis`)

```bash
POSTGRES_USER=username
POSTGRES_PASS=password
POSTGRES_HOST=0.0.0.0
POSTGRES_PORT=5432
POSTGRES_DNAME=postgis
```

Or

- **DATABASE_URL** (str): Full postgres DSN in form of `postgresql://user[:password]@[host][:port][/dbname]`

```bash
DATABASE_URL=postgresql://username:password@0.0.0.0:5432/postgis
```

#### Connection Pools configuration

- **DB_MIN_CONN_SIZE** (int): Number of connection the pool will be initialized with. Default is `1`
- **DB_MAX_CONN_SIZE** (int): Max number of connections in the pool. Default is `10`
- **DB_MAX_QUERIES** (int): Number of queries after a connection is closed and replaced with a new connection. Default is `50000`
- **DB_MAX_INACTIVE_CONN_LIFETIME** (float): Number of seconds after which inactive connections in the pool will be closed. Pass 0 to disable this mechanism. Default is `300`

ref: https://magicstack.github.io/asyncpg/current/api/index.html#connection-pools


## DB visibility settings

class: `tipg.settings.DatabaseSettings`

prefix: **`TIPG_DB_`**

- **SCHEMAS** (list of string): Named schemas, `tipg` can look for `Tables` or `Functions`. Default is `["public"]`

#### `Tables`

- **TABLES** (list of string): Allow only specific `Tables`
- **EXCLUDE_TABLES** (list of string): Excludes specific `Tables`
- **EXCLUDE_TABLE_SCHEMAS** (list of string): Excludes `Schemas` for `Tables` lookup.
- **ONLY_SPATIAL_TABLES** (bool): Accept `Tables` without `geometry` or `geography` column. Default is `True`.

#### `Functions`

- **FUNCTIONS** (list of string): Allow only specific `Functions`
- **EXCLUDE_FUNCTIONS** (list of string): Excludes specific `Functions`
- **EXCLUDE_FUNCTION_SCHEMAS** (list of string): Excludes `Schemas` for `Functions` lookup.

```bash
TIPG_DB_SCHEMAS='["myschema", "public"]'
TIPG_DB_EXCLUDE_FUNCTION_SCHEMAS='["public"]'
```

## Table settings

class: `tipg.settings.TableSettings`

prefix: **`TIPG_`**

- **DATETIME_EXTENT** (bool): Fetch datetime extent by going throught all rows. Default is `True`
- **FALLBACK_KEY_NAMES** (list of string): Primary Key names to look for in the tables. Default is `["ogc_fid", "id", "pkey", "gid"]`
- **TABLE_CONFIG** (dict of `TableConfig`)
    - **TABLE_CONFIG_ _ {schemaId}_{tableId} _ _GEOMCOL** (str): Table's geometry/geography column name
    - **TABLE_CONFIG_ _ {schemaId}_{tableId} _ _DATETIMECOL** (str): Table's datetime column name
    - **TABLE_CONFIG_ _ {schemaId}_{tableId} _ _PK** (str): Table's primary key
    - **TABLE_CONFIG_ _ {schemaId}_{tableId} _ _PROPERTIES** (list of string): Select specific properties from table (for filtering and output)

```bash
TIPG_TABLE_CONFIG__pgstac_items__PK=id
```

## Custom SQL Function

`TiPG` allows for exposing custom `SQL Functions` as collections.

`SQL Functions` are any procedural functions defined in the database that match the following criteria.
  - Must be defined to return "SETOF"
  - Functions defined to return "RECORD" must include typed OUT definitions in the function signature
  - All arguments (IN and OUT) must be named
  - Functions that return a geometry will be usable as Features and Vector Tiles, those that do not will be available to return json/csv
  - Functions that match these qualifications will be found based on the DB visibility settings

`SQL Function` arguments will be exposed to the items API as query parameters. Any argument that does not have a default will be required and will return an error if not set as a query parameter. If a function is defined to have z, x, or y parameters, those will be filled from the path parameters in requests to Tile endpoints. If a function has a bounds parameter, that will be populated by the Tile bounding box in requests to Tile endpoints. All other arguments will be filled from query parameters matching the name of the argument and cast to the appropriate types.


### Example
```SQL
CREATE FUNCTION hexagons(
    IN size int DEFAULT 10,
    IN bounds geometry DEFAULT 'srid=4326;POLYGON((-180 -90,-180 90,180 90,180 -90,-180 -90))'::geometry,
    OUT geom geometry,
    OUT i integer,
    OUT j integer
) RETURNS SETOF RECORD AS $$
    SELECT * FROM st_hexagongrid(size, bounds);
$$ LANGUAGE SQL IMMUTABLE PARALLEL SAFE;
```

`TiPG` also accepts externally defined `SQL Functions` that are defined by .sql files in a directory defined by the `TIPG_CUSTOM_SQL_DIRECTORY` configuration setting.

class: `tipg.settings.CustomSQLSettings`

prefix: **`TIPG_`**

- **CUSTOM_SQL_DIRECTORY** (str): path to the directory containing `.sql` files.

```bash
TIPG_CUSTOM_SQL_DIRECTORY=/tmp/functions
```

Note: On each `connection` creation, `TiPG` will look for all the files within the directory and *execute* them within the `pg_temp` schema.

## Vector Tiles settings

class: `tipg.settings.MVTSettings`

prefix: **`TIPG_`**

- **TILE_RESOLUTION** (int): Default is `4096`
- **TILE_BUFFER** (int): Default is `256`
- **TILE_CLIP** (bool): Default is `True`
- **MAX_FEATURE_PER_TILE** (int): Default is `10000`
- **SET_MVT_LAYERNAME** (bool): Set MVT layername from Table ID. Default is `False` ("default")

```bash
TIPG_MAX_FEATURE_PER_TILE=1000
```

## Features settings

class: `tipg.settings.FeaturesSettings`

prefix: **`TIPG_`**

- **DEFAULT_FEATURES_LIMIT** (int): Set the default `Limit` values for `/items` endpoint. Default is `10`
- **MAX_FEATURES_PER_QUERY** (int): Set the maximum number of features the `/items` endpoint can return. Default is `10000`.

```bash
TIPG_DEFAULT_FEATURES_LIMIT=1000 TIPG_MAX_FEATURES_PER_QUERY=2000
```

## Tile Matrix Sets setting

class: `tipg.settings.TMSSettings`

prefix: **`TIPG_`**

- **DEFAULT_TMS** (str): Default TileMatrixSetId for tiles endpoint. Default is `WebMercatorQuad`
- **DEFAULT_MINZOOM** (int): Default `MinZoom` for tiles endpoints. Default is `0`
- **DEFAULT_MAXZOOM** (int): Default `MaxZoom` for tiles endpoints. Default is `22`

## TiPG Application settings

Settings to control `TiPG` default's FastAPI application.

class: `tipg.settings.APISettings`

prefix: **`TIPG_`**

- **NAME** (str): Set custom name for `TiPG` app. Default is `TiPg: OGC Features and Tiles API`
- **DEBUG** (bool): Default is `False`
- **CORS_ORIGIN** (str): Default is `*`
- **CACHECONTROL** (str): Default is `public, max-age=3600`
- **TEMPLATE_DIRECTORY** (str): Path to custom template directory to overwrite the HTML files.
- **ADD_TILES_VIEWER** (bool): Defaults is `True`


CHANGES.md
# Release Notes

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

Note: Minor version `0.X.0` update might break the API, It's recommended to pin `tipg` to minor version: `tipg>=0.1,<0.2`

## [unreleased]

## [1.2.1] - 2025-08-26

* update `starlette-cramjam` requirement to `>=0.4,<0.6`

## [1.2.0] - 2025-07-23

* remove `request: Request` in `Collection.features` and `Collection.get_tile` abstract methods

* pass a `asyncpg.Connection` to `PgCollection.features` and `PgCollection.get_tile` instead of a `Request` object

* switch `pg_get_collection_index` input from `asyncpg.BuildPgPool` to `asyncpg.Connection` object

## [1.1.2] - 2025-07-02

* fix `rel` values for tiling scheme links

## [1.1.1] - 2025-06-13

* remove `tipg_schema` alias in `DatabaseSettings` to support env variable settings

## [1.1.0] - 2025-05-06

* update geojson-pydantic requirement to `>=1.0,<3.0`

## [1.0.1] - 2025-03-17

* fix typo when using catalog_ttl

## [1.0.0] - 2025-03-07

### Added

* add `tilesets` and `viewer` links in `/collections` and `/collections/{collectionId}` response links
* add the ability to use a schema other than pg_temp for installing catalog functions (using `TIPG_DB_APPLICATION_SCHEMA` environment variable)
* re-use pg *connection* for `features_count` and `features`
* add `tipg.collections.Collection` abstract base class

### Changed

* rename `tipg.collections.Collection -> tipg.collections.PgCollection` **breaking change**
* update `Collection.get_tile()` and `Collection.features()` signature to expect `request: Request` as first parameter **breaking change**
* rename `tipg.collections.get_collection_index` to `tipg.collections.pg_get_collection_index` and change the function to use `DatabaseSettings` instance directly instead of keyword option **breaking change**
* update `tipg.collections.pg_get_collection_index` to return a list of PgCollection instead of a Catalog **breaking change**
* update `tipg.collections.register_collection_catalog` to pass `db_settings` to `pg_get_collection_index` function **breaking change**
* remove **deprecated** tiles endpoint with default TileMatrixSet
* rename tilejson endpoint from `/collections/{collectionId}/{tileMatrixSetId}/tilejson.json` to `/collections/{collectionId}/tiles/{tileMatrixSetId}/tilejson.json` **breaking change**
* rename stylejson endpoint from `/collections/{collectionId}/{tileMatrixSetId}/style.json` to `/collections/{collectionId}/tiles/{tileMatrixSetId}/style.json` **breaking change**
* change `database.connect_to_db` input order **breaking change**

    ```python
    # Before
    async def connect_to_db(
        app: FastAPI,
        settings: Optional[PostgresSettings] = None,
        schemas: Optional[List[str]] = None,
        user_sql_files: Optional[List[pathlib.Path]] = None,
        **kwargs,
    ) -> None:

    # Now
    async def connect_to_db(
        app: FastAPI,
        *,
        schemas: List[str],
        tipg_schema: str = "pg_temp",
        user_sql_files: Optional[List[pathlib.Path]] = None,
        settings: Optional[PostgresSettings] = None,
        **kwargs,
    ) -> None:
    ```

### Fixed

* fix URL in HTML templates when behind proxy

## [0.10.1] - 2025-03-04

* remove `PostgresSettings` initialization from main.py

## [0.10.0] - 2025-02-20

* convert tile bbox into collection's CRS for `MVT` where selection (author @callsumzg, https://github.com/developmentseed/tipg/pull/205)

## [0.9.0] - 2025-01-17

* fix serialization of UUID columns (author @giorgiobasile, https://github.com/developmentseed/tipg/pull/199)

* Unify Docker images (deprecate `tipg-uvicorn`)

* Remove `python3.8` support

* Add `python3.13` support

## [0.8.0] - 2024-10-17

* update `starlette-cramjam` dependency and set compression-level default to 6
* update `morecantile` dependency requirement to allow `6.x`

## [0.7.3] - 2024-09-16

* add `TIPG_SORT_COLUMNS` settings to enable/disable columns sorting (default to `True`) (author @mattdiez-at, https://github.com/developmentseed/tipg/pull/187)
* fix `protomaps-leaflet` version in HTML template (co-author @bmcandr, https://github.com/developmentseed/tipg/pull/190)

## [0.7.2] - 2024-08-27

* move back to `fastapi` dependency

## [0.7.1] - 2024-05-07

* move to `fastapi-slim` to avoid unwanted dependencies

## [0.7.0] - 2024-04-01

- update leaflet version
- add `templated=True` in template URL links
- add `(Template URL)` in template URL links title
- remove *deserialization* in `tipg.factory.create_html_response` function
- add `title` option to `create_html_response` method, in order to set the web page title
- add `**kwargs` to `create_html_response` method to allow custom object to be passed to the template
- fix url/path passed to the HTML template
- fix HTML templates when passing Query Parameters
- replace `maplibre` with `leaflet+protomaps-leaflet` to support other TMS in the HTML `viewer`
- deprecate non-tilematrixset prefixed endpoints for tiles/tilejson/stylejson/viewer to better match with the OGC Spec **breaking**
- re-order endpoints
- remove **Z** limit to support negative values

## [0.6.3] - 2024-02-02

- update pydantic's `Field` usage to avoid 3.0 deprecation
- update starlette's `TemplateResponse' usage to avoid deprecation

## [0.6.2] - 2024-01-19

- add `root_path` API settings
- fix invalid `url` parsing in HTML responses

## [0.6.1] - 2024-01-11

- use `spatial_extent` and `datetime_extent` configuration options in `CatalogUpdateMiddleware` and `/refresh` endpoint (author @hrodmn, https://github.com/developmentseed/tipg/pull/164)

## [0.6.0] - 2024-01-09

- update FastAPI version lower limit to `>0.107.0` and adapt for new starlette version
- fix invalid streaming response formatting
- refactor internal table properties handling
- fix sub-model Table settings (https://github.com/developmentseed/tipg/issues/154)

## [0.5.8] - 2024-01-10


tipg/settings.py
"""tipg config."""

import json
import pathlib
from typing import Any, Dict, List, Optional

from pydantic import (
    BaseModel,
    DirectoryPath,
    Field,
    PostgresDsn,
    ValidationInfo,
    field_validator,
    model_validator,
)
from pydantic_settings import BaseSettings


class APISettings(BaseSettings):
    """API settings"""

    name: str = "TiPg: OGC Features and Tiles API"
    debug: bool = False
    cors_origins: str = "*"
    cachecontrol: str = "public, max-age=3600"
    template_directory: Optional[str] = None
    root_path: str = ""

    add_tiles_viewer: bool = True

    catalog_ttl: int = 300

    model_config = {"env_prefix": "TIPG_", "env_file": ".env", "extra": "ignore"}

    @field_validator("cors_origins")
    def parse_cors_origin(cls, v):
        """Parse CORS origins."""
        return [origin.strip() for origin in v.split(",")]


class TableConfig(BaseModel):
    """Configuration to add table options with env variables."""

    geomcol: Optional[str] = None
    datetimecol: Optional[str] = None
    pk: Optional[str] = None
    properties: Optional[List[str]] = None

    model_config = {"extra": "ignore"}

    @field_validator("properties", mode="before")
    def _properties(cls, v: Any) -> Any:
        """set geometry from geo interface or input"""
        if isinstance(v, str):
            return json.loads(v)
        else:
            return v


class TableSettings(BaseSettings):
    """Table configuration settings"""

    fallback_key_names: List[str] = ["ogc_fid", "id", "pkey", "gid"]
    table_config: Dict[str, TableConfig] = {}
    sort_columns: bool = True

    model_config = {
        "env_prefix": "TIPG_",
        "env_file": ".env",
        "env_nested_delimiter": "__",
        "extra": "ignore",
    }


class TMSSettings(BaseSettings):
    """TiPG TMS settings"""

    default_tms: str = "WebMercatorQuad"
    default_minzoom: int = 0
    default_maxzoom: int = 22

    model_config = {"env_prefix": "TIPG_", "env_file": ".env", "extra": "ignore"}


class FeaturesSettings(BaseSettings):
    """TiPG Items settings"""

    default_features_limit: int = Field(10, ge=0)
    max_features_per_query: int = Field(10000, ge=0)

    model_config = {"env_prefix": "TIPG_", "env_file": ".env", "extra": "ignore"}

    @model_validator(mode="after")
    def max_default(self):
        """Set default bounds and srid when this is a function."""
        if self.default_features_limit > self.max_features_per_query:
            raise ValueError(
                f"Invalid combination of `limit` ({self.default_features_limit}) and `max features per query` ({self.max_features_per_query}) values"
            )

        return self


class MVTSettings(BaseSettings):
    """TiPG MVT settings"""

    tile_resolution: int = 4096
    tile_buffer: int = 256
    tile_clip: bool = True
    max_features_per_tile: int = 10000

    set_mvt_layername: Optional[bool] = None

    model_config = {"env_prefix": "TIPG_", "env_file": ".env", "extra": "ignore"}


class PostgresSettings(BaseSettings):
    """Postgres connection settings.

    Attributes:
        postgres_user: postgres username.
        postgres_pass: postgres password.
        postgres_host: hostname for the connection.
        postgres_port: database port.
        postgres_dbname: database name.

    """

    postgres_user: Optional[str] = None
    postgres_pass: Optional[str] = None
    postgres_host: Optional[str] = None
    postgres_port: Optional[int] = None
    postgres_dbname: Optional[str] = None

    database_url: Optional[PostgresDsn] = None

    db_min_conn_size: int = 1
    db_max_conn_size: int = 10
    db_max_queries: int = 50000
    db_max_inactive_conn_lifetime: float = 300

    model_config = {"env_file": ".env", "extra": "ignore"}

    # https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/core/config.py#L42
    @field_validator("database_url", mode="before")
    def assemble_db_connection(
        cls, v: Optional[str], info: ValidationInfo
    ) -> PostgresDsn:
        """Validate db url settings."""
        if isinstance(v, str):
            return PostgresDsn(v)

        return PostgresDsn.build(
            scheme="postgresql",
            username=info.data.get("postgres_user"),
            password=info.data.get("postgres_pass"),
            host=info.data.get("postgres_host", ""),
            port=info.data.get("postgres_port", 5432),
            path=info.data.get("postgres_dbname", ""),
        )


class DatabaseSettings(BaseSettings):
    """TiPg Database settings."""

    schemas: List[str] = ["public"]
    application_schema: str = "pg_temp"
    tables: Optional[List[str]] = None
    exclude_tables: Optional[List[str]] = None
    exclude_table_schemas: Optional[List[str]] = None
    functions: Optional[List[str]] = None
    exclude_functions: Optional[List[str]] = None
    exclude_function_schemas: Optional[List[str]] = None
    datetime_extent: bool = True
    spatial_extent: bool = True

    only_spatial_tables: bool = True

    model_config = {"env_prefix": "TIPG_DB_", "env_file": ".env", "extra": "ignore"}

    @property
    def tipg_schema(self):
        """Compat: use application_schema for input and tipg_schema for attribute.
        see https://github.com/developmentseed/tipg/issues/224
        """
        return self.application_schema


class CustomSQLSettings(BaseSettings):
    """TiPg Custom SQL settings."""

    custom_sql_directory: Optional[DirectoryPath] = None

    model_config = {"env_prefix": "TIPG_", "env_file": ".env", "extra": "ignore"}

    @property
    def sql_files(self) -> Optional[List[pathlib.Path]]:
        """return a list of SQL files within the custom sql directory."""
        if self.custom_sql_directory:
            return list(self.custom_sql_directory.glob("*.sql"))

        return None


class PMTilesSettings(BaseSettings):
    """TiPg PMTiles settings."""

    paths: Optional[List[str]] = None
    sources: Optional[Dict[str, str]] = None

    model_config = {
        "env_prefix": "TIPG_PMTILES",
        "env_file": ".env",
        "env_nested_delimiter": "__",
        "extra": "ignore",
    }
